Wurstpack/wurstscript/lib/systems/Knockback3.wurst

```diff
@@ -0,0 +1,217 @@
+    // parameters, respectively. Example:
+
+    //     import Knocback3
+    //     ...
+    //     unit u = ...
+    //     Knockback3.add(u, 1000., angle(0.), angle(bj_PI / 4.))
+    */
+    static function add(unit u, real velocity, angle groundAngle, angle airAngle)
+        let instVel = velocity*ANIMATION_PERIOD
+        let v       = vec3(0., 0., 0.).polarProject(instVel, groundAngle, airAngle)
+
+        Knockback3 knockback
+        if unitNodes.has(u)
+            knockback      = unitNodes.get(u)
+            knockback.del += v
+        else
+            knockback     = new Knockback3()
+            knockback.u   = u
+            knockback.del = v
+            if size == 1
+                clock.startPeriodic(ANIMATION_PERIOD, function tick)
+
+            if knockback.u.addAbility(HEIGHT_ENABLER)
+                knockback.u.removeAbility(HEIGHT_ENABLER)
+
+
+    static function setVel(unit u, real velocity, angle groundAngle, angle airAngle)
+        if unitNodes.has(u)
+            let knockback = unitNodes.get(u)
+            let instVel   = velocity*ANIMATION_PERIOD
+
+            knockback.del = vec3(0., 0., 0.).polarProject(instVel, groundAngle, airAngle)
+        else
+            add(u, velocity, groundAngle, airAngle)
+
+
+    static function getVel(unit u) returns real
+        var v = 0.
+        if unitNodes.has(u)
+            let knockback = unitNodes.get(u)
+            v             = knockback.del.length()
+
+        return v
+
+
+    // Instance Variables
+    private unit u
+    private vec3 del
+
+
+    private static function tickNearGround(Knockback3 knockback, vec3 newPos3, vec3 pos3, bool newInMap, real velXySquared)
+        if isTerrainWalkable(newPos3.x, newPos3.y) and newInMap
+            knockback.u.setXY(newPos3)
+
+            if knockback.del.z <= isAirborneThreshold
+                knockback.del *= (1. - frictionCoefficientGround)
+
+                if velXySquared > frictionFxThreshold*ANIMATION_PERIOD
+                    addEffect(frictionFxPath, pos3).destr()
+
+            if USE_MOVE_SPEED_MODIFIERS
+                knockback.u.setMoveSpeed(GetUnitDefaultMoveSpeed(knockback.u))
+
+            if USE_PROP_WINDOW_MODIFIERS
+                knockback.u.setPropWindow(GetUnitDefaultPropWindow(knockback.u) * bj_DEGTORAD)
+
+        else
+            knockback.del = knockback.del.project(vec3(0., 0., 1.))
+
+        if knockback.del.z < elasticityThreshold*ANIMATION_PERIOD
+            knockback.del.z = knockback.del.z*-1.*restitutionCoefficientGround
+
+        if knockback.del.z > airborneThreshold*ANIMATION_PERIOD
+            knockback.u.setFlyHeight(pos3.z + knockback.del.z, 0.)
+            knockback.del.z = knockback.del.z - gravity*ANIMATION_PERIOD
+
+
+    private static function tickAboveGround(Knockback3 knockback, vec3 newPos3, vec3 pos3)
+        knockback.del.z -= gravity*ANIMATION_PERIOD
+        let heightDifference = getTerrainZ(newPos3.x, newPos3.y) - getTerrainZ(pos3.x, pos3.y)
```

Comment:
@Frotty it just moves a location and reads data from heightmap. Afaik this is very cheap, so I'm kind of questioning the value of caching it. How confident are you that this is worth it?